### ISSUE TITLE
multi root support?

### ISSUE THREAD

**Issue by lucywang000** [clojure-lsp/clojure-lsp#280]
I noticed that python's [pyright lsp](https://github.com/emacs-lsp/lsp-pyright) only fires up one instance for multiple python projects, would it make sense to have a similar feature in clojure-lsp? 

I used to open lots (5~10) of clojure/script projects, for e.g reading 3rdparty libs code. It doesn't make sense to have 10 lsp instances running and take 10~20GB memory ...

**Comment by ericdallo** [comment]
Yeah, it sound a good idea, but I'm not sure this works for most of clients, including `lsp-mode` (lsp-clojure.el), it'd need some extra work on their side, and I'm no sure what need to be done on server side too

**Comment by bpringe** [comment]
I was thinking about this recently. I think this is a good idea, though I know it might take some time and thought to implement well.

**Comment by bpringe** [comment]
Another thing to consider is that _if_ the memory usage of clojure-lsp can be drastically reduced via graalvm and other methods, this becomes less of a problem. I still think it's a worthwhile thing to do in any case, though.

**Comment by ericdallo** [comment]
Yeah, a one time instance of clojure-lsp would really be a good for performance but probably would need work a huge work from both server and clients (Vscode, Emacs, vim), and make it retro compatible could be even harder I guess
@snoe do you have any suggestions?

**Comment by snoe** [comment]
No suggestions... Here's some random perspectives/ranting. 

I easily run 5 or 6 instances of clojure-lsp at once. But I will shut down instances when I need to. I only actually work on one project at a time. What are our minimum specs?

I know some workplaces do multi repo/multi project/multi language stuff but it should be understood there's costs there, particularly around tooling support and it should be clear that those places should be investing in their tools.

I think that comparing ourselves to things like intellij or vscode backed by huge companies and hundreds of devs is a losing proposition. 

So, in the end, my view is that collaboration and PRs can be welcome, but understand there's a high effort to maintain compatibility and performance and stability across servers and clients and to do the testing across those environments.

**Comment by starball5** [comment]
Related Stack Overflow question: [How to open multiple projects in VSCode with only one LSP connection?](https://stackoverflow.com/q/75451519/11107541)

**Comment by bpringe** [comment]
> I have multiple projects (git-repos). I have opened these in one multi-root workspace using Add Folder to Workspace.... Since these are all Clojure projects, and I have Calva installed, VSCode identified the correct LSP server clojure-lsp. But VS Code starts one LSP server for each folder/project in my multi-root workspace. I want only one.

> How can I tell VS Code to start only one LSP server for all folders in my workspace?


@ericdallo Isn't there a way to use a single deps.edn file with local references to other projects to get clojure-lsp to support multiple projects with one instance? I seem to recall @seancorfield saying he did something like that.

**Comment by witek** [comment]
> Isn't there a way to use a single deps.edn file with local references to other projects to get clojure-lsp to support multiple projects with one instance? I seem to recall @seancorfield saying he did something like that.

You can configure clojure-lsp in this way. But your IDE remains the problem. It starts a new `clojure-lsp` for each project. With Emacs this can be hacked in different ways. For example you can override `lsp-suggest-project-root` to point multiple projects into one and the same project. This solution does not interfere with other packages (magit for example).

```
(defun lsp--suggest-project-root ()
    "Get project root."
    (let ((dir (project-root (project-current))))
      (if (or (string-prefix-p "/p/my-project-1/" dir)
              (string-prefix-p "/p/my-lib-1/" dir)
              (string-prefix-p "/p/my-lib-2/" dir))
          (project-root (project-try-vc "/p/my-lsp-project/"))
        (project-root (project-try-vc dir)))
      ))
```

**Comment by bpringe** [comment]
I get the impression that the way Sean structured one of his projects, only one clojure-lsp process was started, which served all projects in the one big project. You'd have to get more info from him though. Here's a related Slack thread: https://clojurians.slack.com/archives/CBE668G4R/p1674579601969449?thread_ts=1674566844.014189&cid=CBE668G4R.

**Comment by seancorfield** [comment]
Right, it's a monorepo -- one big Clojure project -- containing ~160 "components" which each have their own `deps.edn` but those are irrelevant for dev/REPL usage. I've blogged extensively about this at corfield.org in the series of monorepo posts over the years.

**Comment by ericdallo** [comment]
Correct @bpringe, clojure-lsp project itself is a mono-repo deps.edn project with a deps.edn at the root pointing to other subprojects, I work all the time with one clojure-lsp session running even on emacs @witek, probably because doom-emacs has `lsp-auto-guess-root` disabled.
It's the same idea of polylith and works pretty good, I intend to add a tutorial on clojure-lsp webpage explaining a little more

**Comment by seancorfield** [comment]
Where are we on this issue? It feels like quite a bit of work has been done on multi-root workspaces in Calva (which doesn't help Emacs users), but is there something to resolve for LSP at the core here, or is this an editor/client issue?

**Comment by ericdallo** [comment]
ATM clojure-lsp doesn't support having multiple projects under the same process and would be a vast refactor to support that, fortunately, for mono-repo projects the suggestion is open the project root, clojure-lsp works well even for huge projects, and having all sub projects under the same process works (clojure-lsp project itself is a mono-repo project).
Regarding opening process when opening 3th party libs, that should not happen, the lsp client should use the same running workspace to open 3th libs (Emacs lsp-mode use to have that issue, but was fixed a long time ago)

**Comment by seancorfield** [comment]
So... this issue should be closed?

**Comment by ericdallo** [comment]
It's a capability not supported right now, and there are alternatives, but still, it's something nice to have in the future, so I think we can keep it open, not a priority right now though.
